File: src/robinhood_options_mobile/lib/model/group_performance_analytics.dart
/// Member performance metrics for leaderboard and analytics
class MemberPerformanceMetrics {
  final String memberId;
  final String memberName;
  final String? memberPhotoUrl;
  final double totalReturnPercent;
  final double totalReturnDollars;
  final double winRate; // Percentage (0-100)
  final int totalTrades;
  final int winningTrades;
  final int losingTrades;
  final double averageWin;
  final double averageLoss;
  final double profitFactor;
  final double sharpeRatio;
  final double maxDrawdownPercent;
  final double? avgHoldTimeHours;
  final DateTime? firstTradeDate;
  final DateTime? lastTradeDate;

  MemberPerformanceMetrics({
    required this.memberId,
    required this.memberName,
    this.memberPhotoUrl,
    required this.totalReturnPercent,
    required this.totalReturnDollars,
    required this.winRate,
    required this.totalTrades,
    required this.winningTrades,
    required this.losingTrades,
    required this.averageWin,
    required this.averageLoss,
    required this.profitFactor,
    required this.sharpeRatio,
    required this.maxDrawdownPercent,
    this.avgHoldTimeHours,
    this.firstTradeDate,
    this.lastTradeDate,
  });

  MemberPerformanceMetrics.fromJson(Map<String, dynamic> json)
      : memberId = json['memberId'] as String,
        memberName = json['memberName'] as String,
        memberPhotoUrl = json['memberPhotoUrl'] as String?,
        totalReturnPercent = json['totalReturnPercent'] as double? ?? 0,
        totalReturnDollars = json['totalReturnDollars'] as double? ?? 0,
        winRate = json['winRate'] as double? ?? 0,
        totalTrades = json['totalTrades'] as int? ?? 0,
        winningTrades = json['winningTrades'] as int? ?? 0,
        losingTrades = json['losingTrades'] as int? ?? 0,
        averageWin = json['averageWin'] as double? ?? 0,
        averageLoss = json['averageLoss'] as double? ?? 0,
        profitFactor = json['profitFactor'] as double? ?? 0,
        sharpeRatio = json['sharpeRatio'] as double? ?? 0,
        maxDrawdownPercent = json['maxDrawdownPercent'] as double? ?? 0,
        avgHoldTimeHours = json['avgHoldTimeHours'] as double?,
        firstTradeDate = json['firstTradeDate'] != null
            ? DateTime.parse(json['firstTradeDate'] as String)
            : null,
        lastTradeDate = json['lastTradeDate'] != null
            ? DateTime.parse(json['lastTradeDate'] as String)
            : null;

  Map<String, dynamic> toJson() {
    return {
      'memberId': memberId,
      'memberName': memberName,
      'memberPhotoUrl': memberPhotoUrl,
      'totalReturnPercent': totalReturnPercent,
      'totalReturnDollars': totalReturnDollars,
      'winRate': winRate,
      'totalTrades': totalTrades,
      'winningTrades': winningTrades,
      'losingTrades': losingTrades,
      'averageWin': averageWin,
      'averageLoss': averageLoss,
      'profitFactor': profitFactor,
      'sharpeRatio': sharpeRatio,
      'maxDrawdownPercent': maxDrawdownPercent,
      'avgHoldTimeHours': avgHoldTimeHours,
      'firstTradeDate': firstTradeDate?.toIso8601String(),
      'lastTradeDate': lastTradeDate?.toIso8601String(),
    };
  }

  @override
  String toString() =>
      'MemberPerformanceMetrics(memberId: $memberId, memberName: $memberName, totalReturnPercent: $totalReturnPercent, winRate: $winRate)';
}

/// Aggregate group performance metrics
class GroupPerformanceMetrics {
  final String groupId;
  final double groupTotalReturnPercent;
  final double groupTotalReturnDollars;
  final double groupAverageReturnPercent;
  final double groupAverageReturnDollars;
  final int totalMembersTraded;
  final int totalGroupTrades;
  final double groupWinRate;
  final double groupAverageSharpeRatio;
  final double topPerformerReturnPercent;
  final String? topPerformerId;
  final int membersWithPositiveReturn;
  final int membersWithNegativeReturn;
  final DateTime? timeRangeStart;
  final DateTime? timeRangeEnd;

  GroupPerformanceMetrics({
    required this.groupId,
    required this.groupTotalReturnPercent,
    required this.groupTotalReturnDollars,
    required this.groupAverageReturnPercent,
    required this.groupAverageReturnDollars,
    required this.totalMembersTraded,
    required this.totalGroupTrades,
    required this.groupWinRate,
    required this.groupAverageSharpeRatio,
    required this.topPerformerReturnPercent,
    this.topPerformerId,
    required this.membersWithPositiveReturn,
    required this.membersWithNegativeReturn,
    this.timeRangeStart,
    this.timeRangeEnd,
  });

  GroupPerformanceMetrics.fromJson(Map<String, dynamic> json)
      : groupId = json['groupId'] as String,
        groupTotalReturnPercent =
            json['groupTotalReturnPercent'] as double? ?? 0,
        groupTotalReturnDollars =
            json['groupTotalReturnDollars'] as double? ?? 0,
        groupAverageReturnPercent =
            json['groupAverageReturnPercent'] as double? ?? 0,
        groupAverageReturnDollars =
            json['groupAverageReturnDollars'] as double? ?? 0,
        totalMembersTraded = json['totalMembersTraded'] as int? ?? 0,
        totalGroupTrades = json['totalGroupTrades'] as int? ?? 0,
        groupWinRate = json['groupWinRate'] as double? ?? 0,
        groupAverageSharpeRatio =
            json['groupAverageSharpeRatio'] as double? ?? 0,
        topPerformerReturnPercent =
            json['topPerformerReturnPercent'] as double? ?? 0,
        topPerformerId = json['topPerformerId'] as String?,
        membersWithPositiveReturn =
            json['membersWithPositiveReturn'] as int? ?? 0,
        membersWithNegativeReturn =
            json['membersWithNegativeReturn'] as int? ?? 0,
        timeRangeStart = json['timeRangeStart'] != null
            ? DateTime.parse(json['timeRangeStart'] as String)
            : null,
        timeRangeEnd = json['timeRangeEnd'] != null
            ? DateTime.parse(json['timeRangeEnd'] as String)
            : null;

  Map<String, dynamic> toJson() {
    return {
      'groupId': groupId,
      'groupTotalReturnPercent': groupTotalReturnPercent,
      'groupTotalReturnDollars': groupTotalReturnDollars,
      'groupAverageReturnPercent': groupAverageReturnPercent,
      'groupAverageReturnDollars': groupAverageReturnDollars,
      'totalMembersTraded': totalMembersTraded,
      'totalGroupTrades': totalGroupTrades,
      'groupWinRate': groupWinRate,
      'groupAverageSharpeRatio': groupAverageSharpeRatio,
      'topPerformerReturnPercent': topPerformerReturnPercent,
      'topPerformerId': topPerformerId,
      'membersWithPositiveReturn': membersWithPositiveReturn,
      'membersWithNegativeReturn': membersWithNegativeReturn,
      'timeRangeStart': timeRangeStart?.toIso8601String(),
      'timeRangeEnd': timeRangeEnd?.toIso8601String(),
    };
  }

  @override
  String toString() =>
      'GroupPerformanceMetrics(groupId: $groupId, groupTotalReturnPercent: $groupTotalReturnPercent, totalMembersTraded: $totalMembersTraded)';
}

/// Enumeration for time period filters
enum TimePeriodFilter {
  oneWeek,
  oneMonth,
  threeMonths,
  sixMonths,
  oneYear,
  allTime,
}

extension TimePeriodFilterExt on TimePeriodFilter {
  String get displayName {
    switch (this) {
      case TimePeriodFilter.oneWeek:
        return '1 Week';
      case TimePeriodFilter.oneMonth:
        return '1 Month';
      case TimePeriodFilter.threeMonths:
        return '3 Months';
      case TimePeriodFilter.sixMonths:
        return '6 Months';
      case TimePeriodFilter.oneYear:
        return '1 Year';
      case TimePeriodFilter.allTime:
        return 'All Time';
    }
  }

  DateTime? getStartDate() {
    final now = DateTime.now();
    switch (this) {
      case TimePeriodFilter.oneWeek:
        return now.subtract(const Duration(days: 7));
      case TimePeriodFilter.oneMonth:
        return now.subtract(const Duration(days: 30));
      case TimePeriodFilter.threeMonths:
        return now.subtract(const Duration(days: 90));
      case TimePeriodFilter.sixMonths:
        return now.subtract(const Duration(days: 180));
      case TimePeriodFilter.oneYear:
        return now.subtract(const Duration(days: 365));
      case TimePeriodFilter.allTime:
        return null;
    }
  }
}

File: src/robinhood_options_mobile/functions/src/group-performance-analytics.ts
import { onCall, HttpsError } from "firebase-functions/v2/https";
import { getFirestore, Timestamp } from "firebase-admin/firestore";

const db = getFirestore();

type Trade = {
  side?: string;
  state?: string;
  instrument_id?: string;
  cumulative_quantity?: number;
  average_price?: number;
  fees?: number;
  created_at?: Timestamp | Date;
};

const dateValue = (value: unknown): Date | null => {
  if (value instanceof Timestamp) return value.toDate();
  if (value instanceof Date) return value;
  if (typeof value === "string") {
    const parsed = new Date(value);
    return Number.isNaN(parsed.getTime()) ? null : parsed;
  }
  return null;
};

const iso = (value: Date | null): string | null => value?.toISOString() ?? null;

/**
 * Build performance metrics for one group member.
 * @param {string} memberId Member user ID.
 * @param {Date | null} startDate Inclusive start of the reporting period.
 * @param {Date} endDate Inclusive end of the reporting period.
 * @return {Promise<Record<string, unknown>>} Member performance metrics.
 */
async function getMemberMetrics(memberId: string, startDate: Date | null,
  endDate: Date): Promise<Record<string, unknown>> {
  const userSnapshot = await db.collection("user").doc(memberId).get();
  const user = userSnapshot.data() ?? {};
  const orderQuery = db.collection("user").doc(memberId)
    .collection("instrumentOrder").where("created_at", "<=", endDate);
  const orderSnapshot = await orderQuery.orderBy("created_at").get();
  const byInstrument = new Map<string, Trade[]>();

  for (const document of orderSnapshot.docs) {
    const order = document.data() as Trade;
    const instrumentId = order.instrument_id;
    const state = order.state?.toLowerCase();
    if ((state !== "filled" && state !== "executed") ||
      !instrumentId) continue;
    const orders = byInstrument.get(instrumentId) ?? [];
    orders.push(order);
    byInstrument.set(instrumentId, orders);
  }

  let totalTrades = 0;
  let winningTrades = 0;
  let losingTrades = 0;
  let totalReturnDollars = 0;
  let totalWinAmount = 0;
  let totalLossAmount = 0;
  let sumReturns = 0;
  let sumReturnsSquared = 0;
  let totalHoldTime = 0;
  let firstTradeDate: Date | null = null;
  let lastTradeDate: Date | null = null;

  for (const orders of byInstrument.values()) {
    const buyQueue: Array<{
      price: number;
      quantity: number;
      fees: number;
      date: Date | null;
    }> = [];
    for (const order of orders) {
      const quantity = order.cumulative_quantity ?? 0;
      const price = order.average_price ?? 0;
      const fees = order.fees ?? 0;
      const createdAt = dateValue(order.created_at);
      if (!quantity) continue;
      if (createdAt && (!firstTradeDate || createdAt < firstTradeDate)) {
        firstTradeDate = createdAt;
      }
      const side = order.side?.toLowerCase();
      if (side === "buy") {
        buyQueue.push({ price, quantity, fees, date: createdAt });
        continue;
      }
      if (side !== "sell") continue;
      let remaining = quantity;
      while (remaining > 0 && buyQueue.length > 0) {
        const buy = buyQueue[0];
        const matched = Math.min(remaining, buy.quantity);
        const pnl = (price - buy.price) * matched -
          fees * (matched / quantity) - buy.fees * (matched / buy.quantity);
        const isInPeriod = !startDate ||
          (createdAt !== null && createdAt >= startDate);
        if (!isInPeriod) {
          remaining -= matched;
          buy.quantity -= matched;
          if (buy.quantity <= 0) buyQueue.shift();
          continue;
        }
        const costBasis = buy.price * matched;
        const pnlPercent = costBasis > 0 ? (pnl / costBasis) * 100 : 0;
        totalTrades++;
        totalReturnDollars += pnl;
        sumReturns += pnlPercent;
        sumReturnsSquared += pnlPercent * pnlPercent;
        if (pnl > 0) {
          winningTrades++;
          totalWinAmount += pnl;
        } else if (pnl < 0) {
          losingTrades++;
          totalLossAmount += Math.abs(pnl);
        }
        if (buy.date && createdAt) {
          totalHoldTime += Math.floor(
            (createdAt.getTime() - buy.date.getTime()) / (60 * 60 * 1000));
        }
        if (createdAt) lastTradeDate = createdAt;
        remaining -= matched;
        buy.quantity -= matched;
        if (buy.quantity <= 0) buyQueue.shift();
      }
    }
  }

  const totalReturnPercent = totalTrades > 0 ? sumReturns / totalTrades : 0;
  const variance = totalTrades > 1 ?
    (sumReturnsSquared / totalTrades) - totalReturnPercent ** 2 : 0;
  const standardDeviation = variance > 0 ? Math.sqrt(variance) : 0;
  const sharpeRatio = standardDeviation > 0 ?
    (totalReturnPercent / standardDeviation) * Math.sqrt(252) : 0;
  const profitFactor = totalLossAmount > 0 ? totalWinAmount / totalLossAmount :
    0;

  return {
    memberId,
    memberName: user.name ?? "Unknown",
    memberPhotoUrl: user.photoUrl ?? null,
    totalReturnPercent,
    totalReturnDollars,
    winRate: totalTrades > 0 ? (winningTrades / totalTrades) * 100 : 0,
    totalTrades,
    winningTrades,
    losingTrades,
    averageWin: winningTrades > 0 ? totalWinAmount / winningTrades : 0,
    averageLoss: losingTrades > 0 ? totalLossAmount / losingTrades : 0,
    profitFactor,
    sharpeRatio,
    maxDrawdownPercent: totalTrades > 0 && totalLossAmount > 0 ?
      (totalLossAmount / (totalWinAmount + totalLossAmount)) * 100 : 0,
    avgHoldTimeHours: totalTrades > 0 ? totalHoldTime / totalTrades : null,
    firstTradeDate: iso(firstTradeDate),
    lastTradeDate: iso(lastTradeDate),
  };
}

export const getGroupPerformanceAnalytics = onCall(async (request) => {
  if (!request.auth) {
    throw new HttpsError("unauthenticated", "User must be authenticated");
  }
  const groupId = request.data?.groupId;
  if (typeof groupId !== "string" || !groupId) {
    throw new HttpsError("invalid-argument", "groupId is required");
  }
  const groupSnapshot = await db.collection("investor_groups")
    .doc(groupId).get();
  if (!groupSnapshot.exists) {
    throw new HttpsError("not-found", "Group not found");
  }
  const group = groupSnapshot.data() ?? {};
  const members = Array.isArray(group.members) ?
    group.members as string[] : [];
  if (!members.includes(request.auth.uid)) {
    throw new HttpsError(
      "permission-denied", "You must be a member of this group");
  }

  const startDate = request.data.startDate ?
    new Date(request.data.startDate) : null;
  const endDate = request.data.endDate ?
    new Date(request.data.endDate) : new Date();
  const memberMetrics = await Promise.all(
    members.map((memberId) => getMemberMetrics(memberId, startDate, endDate)));
  const traded = memberMetrics.filter((member) =>
    (member.totalTrades as number) > 0);
  const totalTrades = traded.reduce(
    (sum, member) => sum + (member.totalTrades as number), 0);
  const totalWinningTrades = traded.reduce(
    (sum, member) => sum + (member.winningTrades as number), 0);
  const averageReturn = traded.length > 0 ?
    traded.reduce((sum, member) =>
      sum + (member.totalReturnPercent as number), 0) / traded.length : 0;
  const totalDollars = traded.reduce(
    (sum, member) => sum + (member.totalReturnDollars as number), 0);
  const averageSharpe = traded.length > 0 ?
    traded.reduce((sum, member) => sum + (member.sharpeRatio as number), 0) /
    traded.length : 0;
  const top = traded.reduce<Record<string, unknown> | null>((best, member) =>
    !best || (member.totalReturnPercent as number) >
      (best.totalReturnPercent as number) ? member : best, null);

  return {
    groupMetrics: {
      groupId,
      groupTotalReturnPercent: averageReturn,
      groupTotalReturnDollars: totalDollars,
      groupAverageReturnPercent: averageReturn,
      groupAverageReturnDollars: traded.length > 0 ?
        totalDollars / traded.length : 0,
      totalMembersTraded: traded.length,
      totalGroupTrades: totalTrades,
      groupWinRate: totalTrades > 0 ?
        (totalWinningTrades / totalTrades) * 100 : 0,
      groupAverageSharpeRatio: averageSharpe,
      topPerformerReturnPercent: (top?.totalReturnPercent as number) ?? 0,
      topPerformerId: (top?.memberId as string) ?? null,
      membersWithPositiveReturn: traded.filter((m) =>
        (m.totalReturnPercent as number) > 0).length,
      membersWithNegativeReturn: traded.filter((m) =>
        (m.totalReturnPercent as number) < 0).length,
      timeRangeStart: iso(startDate),
      timeRangeEnd: iso(endDate),
    },
    memberMetrics,
  };
});
